libyul/backends/wasm/polyfill/Keccak.yul

```diff
@@ -18,7 +18,332 @@
+	i64.store(i32.add(round_consts, 144:i32), 0x000000000000800A)
+	i64.store(i32.add(round_consts, 152:i32), 0x800000008000000A)
+	i64.store(i32.add(round_consts, 160:i32), 0x8000000080008081)
+	i64.store(i32.add(round_consts, 168:i32), 0x8000000000008080)
+	i64.store(i32.add(round_consts, 176:i32), 0x0000000080000001)
+	i64.store(i32.add(round_consts, 184:i32), 0x8000000080008008)
+	//    const int keccakf_rotc[24] = {
+	//        1,  3,  6,  10, 15, 21, 28, 36, 45, 55, 2,  14,
+	//        27, 41, 56, 8,  25, 43, 62, 18, 39, 61, 20, 44
+	//    };
+	let keccakf_rotc:i32 := keccakf_rotc_offset(context_offset)
+	i64.store(i32.add(keccakf_rotc, 0:i32), bswap64(0x0103060A0F151C24))
+	i64.store(i32.add(keccakf_rotc, 8:i32), bswap64(0x2D37020E1B293808))
+	i64.store(i32.add(keccakf_rotc, 16:i32), bswap64(0x192B3E12273D142C))
+	//    const int keccakf_piln[24] = {
+	//        10, 7,  11, 17, 18, 3, 5,  16, 8,  21, 24, 4,
+	//        15, 23, 19, 13, 12, 2, 20, 14, 22, 9,  6,  1
+	//    };
+	let keccakf_piln:i32 := keccakf_piln_offset(context_offset)
+	i64.store(i32.add(keccakf_piln, 0:i32), bswap64(0x0A070B1112030510))
+	i64.store(i32.add(keccakf_piln, 8:i32), bswap64(0x081518040F17130D))
+	i64.store(i32.add(keccakf_piln, 16:i32), bswap64(0x0C02140E16090601))
+	let bc:i32 := bc_offset(context_offset)
+	memset(i32.add(bc, 0:i32), 0:i32, 40:i32)
+}
+function keccak_update(context_offset:i32, input_offset:i32, input_length:i32) {
+	//	size_t i;
+	//	int j;
+	//	j = c->pt;
+	//	for (i = 0; i < len; i++) {
+	//		c->st.b[j++] ^= ((const uint8_t *) data)[i];
+	//		if (j >= c->rsiz) {
+	//			keccakf(c->st.q);
+	//			j = 0;
+	//		}
+	//	}
+	//	c->pt = j;
+	let j:i32 := i32.wrap_i64(i64.load(pt_offset(context_offset)))
+	for { let i:i32 := 0:i32 } i32.lt_u(i, input_length) { i := i32.add(i, 1:i32) }
+	{
+		let ptr:i32 := i32.add(context_offset, j)
+		j := i32.add(j, 1:i32)
+		i8_store(ptr, i64.xor(i8_load(ptr), i8_load(i32.add(input_offset, i))))
+		if i32.ge_u(j, i32.wrap_i64(i64.load(rsiz_offset(context_offset))))
+		{
+			keccakf(context_offset)
+			j := 0:i32
+		}
+	}
+	i64.store(pt_offset(context_offset), i64.extend_i32_u(j))
+}
+function keccak_finish(context_offset:i32) {
+	// c->st.b[c->pt] |= 0x01; // for keccak <- we want to use this.
+	let ptr:i32 := i32.add(context_offset, i32.wrap_i64(i64.load(pt_offset(context_offset))))
+	i8_store(ptr, i64.or(i8_load(ptr), 0x01))
+	// c->st.b[c->rsiz - 1] ^= 0x80
+	ptr := i32.add(context_offset, i32.sub(i32.wrap_i64(i64.load(rsiz_offset(context_offset))), 1:i32))
+	i8_store(ptr, i64.xor(i8_load(ptr), 0x80))
+	// keccakf(c->st.q);
+	keccakf(context_offset)
+}
 function keccak256(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
-	// TODO implement
-	unreachable()
+	let input_offset:i32 := u256_to_i32(x1, x2, x3, x4)
+	let input_length:i32 := u256_to_i32(y1, y2, y3, y4)
+//	debug.print32(input_length)
+//	debug.print32(input_offset)
+//	debug.print64(i64.load(input_offset))
```

Comment:
_(copied from #9830)_
@axic the data where `input_offset` is pointing to is always zero. I'm not yet sure why this happens.

```
[0] DEBUG print32(0x60)
[0] DEBUG print32(0xa0)
[0] DEBUG print64(0x0)
[0] DEBUG print64(0x0)
[0] DEBUG print64(0x0)
[0] DEBUG print64(0x0)
[0] DEBUG print64(0x0)
[0] DEBUG print64(0x0)
[0] DEBUG print64(0x0)
[0] DEBUG print64(0x0)
```

Reply:
This is the only missing piece to have a fully functional version. Somehow it looks like that the data is not correctly copied to the `input` location. The memory where `input offset` is pointing to contains only zeros. 

If I setup the `input` manually, e.g. with
```
let input_offset:i32 := 0xE000:i32
i64.store(input_offset, 0x6f6f66) // -> "foo"
let input_length:i32 := 3:i32
```
the keccak function seem to return the correct hash.